src/provider/mod.rs

```diff
@@ -0,0 +1,181 @@
+use kube::{Api, Client};
+use kubelet::backoff::ExponentialBackoffStrategy;
+use kubelet::log::Sender;
+use kubelet::node::Builder;
+use kubelet::pod::Pod;
+use kubelet::provider::Provider;
+use log::{debug, error};
+
+use crate::provider::error::StackableError;
+use crate::provider::error::StackableError::{CrdMissing, PodValidationError};
+use crate::provider::repository::package::Package;
+use crate::provider::states::download_package::Downloading;
+use crate::provider::states::terminated::Terminated;
+
+pub struct StackableProvider {
+    client: Client,
+    parcel_directory: PathBuf,
+    config_directory: PathBuf,
+    log_directory: PathBuf,
+}
+
+pub const CRDS: &'static [&'static str] = &["repositories.stable.stackable.de"];
+
+mod error;
+mod repository;
+mod states;
+
+pub struct PodState {
+    client: Client,
+    parcel_directory: PathBuf,
+    download_directory: PathBuf,
+    config_directory: PathBuf,
+    log_directory: PathBuf,
+    package_download_backoff_strategy: ExponentialBackoffStrategy,
+    service_name: String,
+    package: Package,
+    process_handle: Option<Child>,
+}
+
+impl PodState {
+    pub fn get_service_config_directory(&self) -> PathBuf {
+        self.config_directory.join(&self.service_name)
+    }
+
+    pub fn get_service_package_directory(&self) -> PathBuf {
+        self.parcel_directory
+            .join(&self.package.get_directory_name())
+    }
+
+    pub fn get_service_log_directory(&self) -> PathBuf {
+        self.log_directory.join(&self.service_name)
+    }
+}
+
+impl StackableProvider {
+    pub async fn new(
+        client: Client,
+        parcel_directory: PathBuf,
+        config_directory: PathBuf,
+        log_directory: PathBuf,
+    ) -> Result<Self, StackableError> {
+        let provider = StackableProvider {
+            client,
+            parcel_directory,
+            config_directory,
+            log_directory,
+        };
+        let missing_crds = provider.check_crds().await;
+        if missing_crds.is_empty() {
+            debug!("All required CRDS present!");
+            return Ok(provider);
+        } else {
+            debug!("Missing required CDRS");
+            return Err(CrdMissing { missing_crds });
+        }
+    }
+
+    fn get_package(&self, pod: &Pod) -> Result<Package, StackableError> {
+        let containers = pod.containers();
+        if containers.len().ne(&1) {
+            let e = PodValidationError {
+                msg: String::from("Size of containers list in PodSpec has to be exactly 1"),
+            };
+            return Err(e);
+        } else {
+            // List has exactly one value, try to parse this
+            if let Ok(Some(reference)) = containers[0].image() {
+                return Package::try_from(reference);
+            } else {
+                let e = PodValidationError {
+                    msg: String::from("Unable to get package reference from pod"),
```

Comment:
We cannot print the reference itself, because not getting one triggered the error, but I've added the pod name at least - since we currently enforce it only having one container and that can only have one image name this should be specific enough to identify exactly which field caused this.